src/planner/binder/expression/bind_star_expression.cpp

```diff
@@ -89,6 +98,102 @@ void Binder::ReplaceStarExpression(unique_ptr<ParsedExpression> &expr, unique_pt
 	    *expr, [&](unique_ptr<ParsedExpression> &child_expr) { ReplaceStarExpression(child_expr, replacement); });
 }
 
+void Binder::ReplaceUnpackedStarExpression(unique_ptr<ParsedExpression> &expr,
+                                           vector<unique_ptr<ParsedExpression>> &replacements,
+                                           ColumnUnpackResult &parent) {
+	D_ASSERT(expr);
+	if (expr->GetExpressionClass() == ExpressionClass::STAR) {
```

Comment:
Getting back into the code, I remember why I went this route
Every input expression can create `n` output expressions, in that sense it's kind of like a table-in-out function

GetChild() gives back the list of expressions that the input expression expanded into, this works like an iterator, forwarding it internally on every call:

```c++
			// Replace children
			vector<unique_ptr<ParsedExpression>> new_children;
			for (auto &unused : function_expr.children) {
				(void)unused;
				auto child_expressions = children.GetChild();
				for (auto &child : child_expressions) {
					new_children.push_back(std::move(child));
				}
			}
			function_expr.children = std::move(new_children);

			// Replace FILTER
			if (function_expr.filter) {
				auto child_expressions = children.GetChild();
				if (child_expressions.size() != 1) {
					throw NotImplementedException("*COLUMNS(...) is not supported in the filter expression");
				}
				function_expr.filter = std::move(child_expressions[0]);
			}
```

This is done because ParsedExpressionIterator has no way to provide me with context as to what expression I am iterating over, is it a projection list or are we looking at a filter?
So here I am iterating through the expressions in the same way that the ParsedExpressionIterator does so I can relate the resulting expansion to the destination.

Reply:
Perhaps I can simplify this if I create an extended version of the ParsedExpressionIterator that not only provides the ParsedExpression but wraps it in a struct that contains the origin of that expression, so I can make replacements directly

Reply:
Hmm actually I don't think so

Let's say we're replacing one child of the children of a FunctionExpression.
We can't insert into this vector directly, that would invalidate the iterator that we're using to iterate through it

This is not a problem if we were only doing a 1->1 replacement, then we can change the unique_ptr without disturbing the vector, but if we replace with 2 or more expressions we can't get away with this.

Reply:
Makes sense - I think we can keep the current implementation then

Reply:
I think we can simplify this still by not doing the unpacking using the `ParsedExpressionIterator`. Essentially, there are two components to this traversal: the actual unpacking (only supported for a limited number of expressions), and traversing into the children. The actual unpacking needs to be special cased (as it already is). The traversal can be done generically using the `ParsedExpressionIterator`. I think we can simplify this code by first doing the unpacking and using the iterator only for the traversal. That way we can get rid of the `ColumnUnpackResult` entirely because we directly unpack into the destination.

Reply:
Yea that simplified the code a ton, thanks for spelling this out, I was looking in the wrong direction